internal/fwserver/server.go

```diff
@@ -9,4 +14,41 @@ import (
 // response type conversions.
 type Server struct {
 	Provider tfsdk.Provider
+
+	// providerSchema is the cached Provider Schema for RPCs that need to
+	// convert configuration data from the protocol. If not found, it will be
+	// fetched from the Provider.GetSchema() method.
+	providerSchema *tfsdk.Schema
+
+	// providerSchemaDiags is the cached Diagnostics obtained while populating
+	// providerSchema. This is to ensure any warnings or errors are also
+	// returned appropriately when fetching providerSchema.
+	providerSchemaDiags diag.Diagnostics
+
+	// providerSchemaMutex is a mutex to protect concurrent providerSchema
+	// access from race conditions.
+	providerSchemaMutex sync.Mutex
+}
+
+// ProviderSchema returns the Schema associated with the Provider. The Schema
+// and Diagnostics are cached on first use.
+func (s *Server) ProviderSchema(ctx context.Context) (*tfsdk.Schema, diag.Diagnostics) {
+	logging.FrameworkTrace(ctx, "Checking ProviderSchema lock")
+	s.providerSchemaMutex.Lock()
+
+	if s.providerSchema != nil {
+		s.providerSchemaMutex.Unlock()
```

Comment:
To ensure future code doesn't break the needs of the mutex to be unlocked, I have seen a slightly different usage pattern (that I like). Something like:

```
s.ProviderSchemaMutex.Lock()
defer s.ProviderSchemaMutex.Unlock()
...
```

This way, whatever branching of the logic we take (the memoized one or not), the locking bit is handled safely and consistently.

Reply:
Thank you for this suggestion, it's a great one and one I'll implement!